### DISCUSSION TITLE
How to log request json body in case of error in Ocelot?

### DISCUSSION THREAD

**Discussion by rajeshgithub001** [ThreeMammals/Ocelot#2012] [Q&A] [answered]
I am serilog to log error from ocelot in case of downstream giving any error.
Using serilog i am sending the exception details to Application Insight.

Serilog configuration:
```json
"Serilog": {
    "Using": [ "Serilog.Sinks.ApplicationInsights" ],
    "MinimumLevel": {
      "Default": "Information",
      "Override": {
        "Microsoft": "Warning"
      }
    },
    "WriteTo": [
      {
        "Name": "ApplicationInsights",
        "Args": {
          "instrumentationKey": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
          "telemetryConverter": "Serilog.Sinks.ApplicationInsights.Sinks.ApplicationInsights.TelemetryConverters.TraceTelemetryConverter, Serilog.Sinks.ApplicationInsights"
        }
      }
    ],
    "Enrich": [ "FromLogContext" ],
    "Properties": {
      "Application": "API GATEWAY"
    }
  }
```
Now i want to log request json body  in case of error only , How will i do it ?

Middleware didn't work as it didn't catch any ocelot exception, so its out of question.

**Comment by MirzaMerdovic** [comment]
Maybe using [DelegatingHandler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html) can solve your problem? 
For example:
```C#
public class FakeHandler : DelegatingHandler
{
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        //do stuff and optionally call the base handler..
        var response = await base.SendAsync(request, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            // TODO: Log stuff ... 
        }

        return response;
    }
}
```

**Reply by raman-m** [reply, marked as answer]
Yes, but Not all errors will be caught by delegating handler :exclamation: 

My recommendation is using `PreErrorResponderMiddleware` injection. See [Middleware Injection](https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html).
It requires analysis of Ocelot error codes, with limited access to exception messages.

### More advanced approach is replacing `IExceptionToErrorMapper` object in DI container.
Just define new class inheriting the `IExceptionToErrorMapper` interface.
Aggregate [HttpExceptionToErrorMapper](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Requester/HttpExceptionToErrorMapper.cs) class member as private one.
Decorate behavior of the class or totally replace by new behavior of Decorator pattern.
So, `public Error Map(Exception exception)` method can be decorated like this:
```csharp
    public class MyExceptionToErrorMapperDecorator : IExceptionToErrorMapper
    {
        private readonly HttpExceptionToErrorMapper _decorated; // New aggregation

        public MyExceptionToErrorMapperDecorator(IServiceProvider serviceProvider)
        {
            _decorated = new HttpExceptionToErrorMapper(serviceProvider);
        }

        public Error Map(Exception exception)
        {
            // Yahoo! This is the place where you can log your lovely exception
            // ... any logging
            return _decorated.Map(exception);
        }
    }
```
Don't forget to replace old mapper in Di container via services collection:
```C#
.ConfigureServices(services => {
        services.RemoveAll<IExceptionToErrorMapper>();
        services.TryAddSingleton<IExceptionToErrorMapper, MyExceptionToErrorMapperDecorator>();
})
```

Hope it helps!
